default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1601 @@
+                                 if get_shipdesign(ID).name(False).startswith(TESTDESIGN_NAME_PART)]
+
+        available_slot_types = {slottype for slotlist in [get_hulltype(hull).slots for hull in available_hulls]
+                                for slottype in slotlist}
+
+        new_parts = [_get_part_type(part) for part in empire.availableShipParts
+                     if part not in self.strictly_worse_parts]
+        pid = self.production_cost.keys()[0]  # TODO: Check if actually location invariant
+        for new_part in new_parts:
+            self.strictly_worse_parts[new_part.name] = []
+            for part_class in ALL_META_CLASSES:
+                if new_part.partClass in part_class:
+                    for old_part in [_get_part_type(part) for part in self.strictly_worse_parts
+                                     if part != new_part.name]:
+                        if old_part.partClass in part_class:
+                            if new_part.capacity >= old_part.capacity:
+                                a = new_part
+                                b = old_part
+                            else:
+                                a = old_part
+                                b = new_part
+                            if (self.production_cost[pid][a.name] <= self.production_cost[pid][b.name]
+                                    and {x for x in a.mountableSlotTypes} >= {x for x in b.mountableSlotTypes}):
+                                # ToDo: Maybe add production_time as additional condition?
+                                self.strictly_worse_parts[a.name].append(b.name)
+                    break
+        available_parts = sorted(self.strictly_worse_parts.keys(),
+                                 key=lambda item: _get_part_type(item).capacity, reverse=True)
+
+        # in case of a load, we need to rebuild our Cache.
+        if not self.testhulls:
+            print "Testhull cache not found. This may happen only at first turn after game start or load."
+            for hullname in available_hulls:
+                des = [des for des in testdesign_names_part if des.endswith(hullname)]
+                if des:
+                    self.testhulls.add(hullname)
+            if verbose:
+                print "Rebuilt Cache. The following hulls are used in testdesigns for parts: ", self.testhulls
+
+        # 1. Update hull test designs
+        print "Updating Testdesigns for hulls..."
+        if verbose:
+            print "Available Hulls: ", available_hulls
+            print "Existing Designs (prefix: %s): " % TESTDESIGN_NAME_HULL,
+            print [x.replace(TESTDESIGN_NAME_HULL, "") for x in testdesign_names_hull]
+        for hull in [get_hulltype(hullname) for hullname in available_hulls
+                     if "%s_%s" % (TESTDESIGN_NAME_HULL, hullname) not in testdesign_names_hull]:
+            num_slots = len([slot for slot in hull.slots])
```

Comment:
`hull.slots` is `fo.ShipSlotVec` and it is implement '__len__` so just `len(hull.slots)` and since it used once just inline it.